cmd/zfs/zfs_main.c

```diff
@@ -6772,12 +6773,31 @@ get_one_dataset(zfs_handle_t *zhp, void *data)
 	}
 
 	/*
-	 * Skip any dataset whos name doesn't have a prefix matching ga_dataset.
+	 * Check if ga_filesystem is set; used for recursive mounting
 	 */
-	if (state->ga_dataset != NULL && strncmp(zfs_get_name(zhp),
-	    state->ga_dataset, strlen(state->ga_dataset)) != 0) {
-		zfs_close(zhp);
-		return (0);
+	if (state->ga_filesystem != NULL) {
+		/*
+		 * Skip any dataset with the canmount=off property.
+		 */
+		if (zfs_prop_get_int(zhp, ZFS_PROP_CANMOUNT) == ZFS_CANMOUNT_OFF) {
+			zfs_close(zhp);
+			return (0);
+		}
+
+		/* Add a trailing / to ga_filesystem */
+		char filesystem[strlen(state->ga_filesystem) + 2];
+		strcpy(filesystem, state->ga_filesystem);
+		strcat(filesystem, "/");
+
+		/*
+		 * Skip any dataset whose name doesn't match ga_filesystem exactly
+		 * or have a prefix matching filesystem.
+		 */
+		if (!strcmp(zfs_get_name(zhp), state->ga_filesystem) == 0 ||
+		    !strncmp(zfs_get_name(zhp), filesystem, strlen(filesystem) == 0)) {
+			zfs_close(zhp);
+			return (0);
```

Comment:
There's no need to do two string comparisons or store the same string with a '/'. Do something similar to this instead (using my recommended rename of the struct field):

```c
const char *fsname = zfs_get_name(zhp);
const size_t parent_len = strlen(state->ga_parent);
boolean_t descendant = true;

if (strncmp(state->ga_parent, fsname, parent_len) == 0) {
	descendant = (fsname[parent_len] == '/' || fsname[parent_len] == '\0');
}

if (!descendant) {
	zfs_close(zhp);
	return (0);
}
```

Reply:
I did this to resolve the following issue:
````
zpool/dataset/name
zpool/dataset/name2
````

The trailing slash is used to ensure we don't accidentally mount `zpool/dataset/name2` wen the target is `zpool/dataset/name`and its children.

Reply:
You should read my test again.

Reply:
Don't know why I didn't see the code block when I made the response.

`fsname[parent_len] ` is definitely a better way to accomplish this, thanks for the suggestion.